tools/codegen/dest/gen_external_aten_fallbacks.py

```diff
@@ -0,0 +1,305 @@
+#
+#   - This function registers external backend kernels, and also generates fallbacks to CPU.
+#     This is useful because pretty much all external backends (e.g. XLA)
+#     do not have full aten coverage.
+#     For operators not implemented by the external backend, our codegen
+#     will register these fallbacks instead.
+#   - Why do we generate fallback for ALL aten ops, including ops that
+#     external backends have already implemented?
+#     Many external backend kernels only work with specific input shapes,
+#     and are written to call into a cpu fallback when given inputs
+#     that they cannot handle.
+@dataclass(frozen=True)
+class GenExternalAtenFallback:
+    target: Union[
+        Literal[Target.NAMESPACED_DEFINITION],
+        Literal[Target.NAMESPACED_DECLARATION],
+        Literal[Target.REGISTRATION],
+    ]
+
+    @method_with_native_function
+    def __call__(self, g: Union[ExternalBackendFunctionsGroup, ExternalBackendFunction]) -> List[str]:
+
+        def gen_out_wrapper(g: ExternalBackendFunctionsGroup) -> Optional[str]:
+            dispatcher_sig = DispatcherSignature.from_schema(g.out.native_function.func)
+            name = dispatcher_sig.name()
+
+            dispatcher_order_args = list(dispatcher.jit_arguments(g.out.native_function.func))
+            tensors = [a for a in dispatcher_order_args if a.type == BaseType(BaseTy.Tensor)]
+            print_args_str = ''.join([f' << " {a.name}=" << {a.name}.toString()' for a in tensors])
+
+            func_name = f'AtenXlaTypeDefault::{name}'
+            return_names = cpp.return_names(g.out.native_function, override_name="x_result")
+            if len(return_names) > 1:
+                updates = '\n  '.join(
+                    f'bridge::XlaUpdateTensors({{{ret_name}}}, {{std::get<{i}>({name}_tmp)}}, {{0}});'
+                    for i, ret_name in enumerate(return_names))
+                returns = f'{dispatcher_sig.returns_type().cpp_type()}({", ".join(return_names)})'
+            else:
+                ret_name = return_names[0]
+                updates = f'bridge::XlaUpdateTensors({{{ret_name}}}, {{{name}_tmp}}, {{0}});'
+                returns = ret_name
+
+            functional_sig = DispatcherSignature.from_schema(g.functional.native_function.func)
+
+            return f"""\
+{dispatcher_sig.defn(name=func_name)} {{
+  XLA_FN_TRACK(3);
+  TF_VLOG(3) << "XLA {name} :"{print_args_str};
+  auto {name}_tmp = AtenXlaType::{functional_sig.name()}({", ".join(a.name for a in functional_sig.arguments())});
+  {updates}
+  return {returns};
+}}
+
+"""
+
+        def gen_unstructured_external(f: ExternalBackendFunction) -> Optional[str]:
+            if not requires_backend_wrapper(f):
+                return None
+
+            def get_device_param(args: List[Argument]) -> str:
+                # TODO: the XLA codegen has specific precedence rules when determining which tensor argument
+                # to use as the device argument.
```

Comment:
cc @wenleix

Reply:
for the XLA case, do they expect all tensors are on the same device? :)

Reply:
Good question, I dunno :) maybe @ailzhang does.

We probably want to keep the logic for choosing which input to get the device from consistent either way though, right?

Reply:
It looks like mixed xla/cpu devices are allowed, at least:
```
>>> import torch_xla.core.xla_model as xm
>>> dev = xm.xla_device()
>>> a = torch.ones(3, device=dev)
>>> b = torch.ones(3)
>>> a + b
tensor([2., 2., 2.], device='xla:0')
>>> a.device
device(type='xla', index=0)
>>> b.device
device(type='cpu')
>>>
```